src/cargo/util/lints.rs

```diff
@@ -0,0 +1,226 @@
+use crate::core::FeatureValue::Dep;
+use crate::core::{Edition, FeatureValue, Package};
+use crate::util::interning::InternedString;
+use crate::{CargoResult, GlobalContext};
+use annotate_snippets::{Level, Renderer, Snippet};
+use cargo_util_schemas::manifest::{TomlLintLevel, TomlToolLints};
+use pathdiff::diff_paths;
+use std::collections::HashSet;
+use std::ops::Range;
+use std::path::Path;
+use toml_edit::ImDocument;
+
+fn get_span(document: &ImDocument<String>, path: &[&str], get_value: bool) -> Option<Range<usize>> {
+    let mut table = document.as_item().as_table_like().unwrap();
+    let mut iter = path.into_iter().peekable();
+    while let Some(key) = iter.next() {
+        let (key, item) = table.get_key_value(key).unwrap();
+        if iter.peek().is_none() {
+            return if get_value {
+                item.span()
+            } else {
+                let leaf_decor = key.dotted_decor();
+                let leaf_prefix_span = leaf_decor.prefix().and_then(|p| p.span());
+                let leaf_suffix_span = leaf_decor.suffix().and_then(|s| s.span());
+                if let (Some(leaf_prefix_span), Some(leaf_suffix_span)) =
+                    (leaf_prefix_span, leaf_suffix_span)
+                {
+                    Some(leaf_prefix_span.start..leaf_suffix_span.end)
+                } else {
+                    key.span()
+                }
+            };
+        }
+        if item.is_table_like() {
+            table = item.as_table_like().unwrap();
+        }
+        if item.is_array() && iter.peek().is_some() {
+            let array = item.as_array().unwrap();
+            let next = iter.next().unwrap();
+            return array.iter().find_map(|item| {
+                if next == &item.to_string() {
+                    item.span()
+                } else {
+                    None
+                }
+            });
+        }
+    }
+    None
+}
+
+/// Gets the relative path to a manifest from the current working directory, or
+/// the absolute path of the manifest if a relative path cannot be constructed
+fn rel_cwd_manifest_path(path: &Path, gctx: &GlobalContext) -> String {
+    diff_paths(path, gctx.cwd())
+        .unwrap_or_else(|| path.to_path_buf())
+        .display()
+        .to_string()
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct LintGroup {
+    pub name: &'static str,
+    pub default_level: LintLevel,
+    pub desc: &'static str,
+    pub edition_lint_opts: Option<(Edition, LintLevel)>,
+}
+
+const RUST_2024_COMPATIBILITY: LintGroup = LintGroup {
+    name: "rust-2024-compatibility",
+    default_level: LintLevel::Allow,
+    desc: "warn about compatibility with Rust 2024",
+    edition_lint_opts: Some((Edition::Edition2024, LintLevel::Deny)),
```

Comment:
We're only making this `Deny` for 2024?  That means people can override this to allow it.  If they do, we should make sure that the dependency truly is unused and doesn't create a feature.